_posts/topology-changes.md

```diff
@@ -0,0 +1,82 @@
+ 
+The third option is to introduce a triage endpoint. We will convert Sales into the triage endpoint and add two more endpoints to handle messages: Sales.PostPaymentProcessing and Sales.Operations, for example. The triage endpoint masquerades the processing endpoints topology to senders. Using a forwarding technique, the triage endpoint inspects incoming messages and then redirects them to the processing endpoint.
+
+The second option is designed to be temporary. We will modify senders to properly route messages to the new endpoint and remove the forwarding logic. On the contrary, the third option comes with the goal of not breaking senders.
+
+### Publish/subscribe caveats
+
+All that presented so far applies well to message sending. Pub/sub is a bit trickier and what we do depends more on the underlying infrastructure. Let's say that we're deploying on AWS (Amazon Web Services), and we are using SQS (Simple Queue Service) and SNS (Simple Notification Service).
+
+Let's imagine that an event is responsible for triggering post-payment operations. Once users pay, the payment processor publishes a `PaymentCompleted` event. Based on that, Salles subscribed to the event and started the post-payment operations. Visually, it's like the following diagram:
+
+<div class="mermaid">
+graph LR
+    P[Publisher] --> T[PaymentCompleted - SNS Topic]
+    T --> S1[Sales - subscription]
+    S1 -->Q1[Sales - SQS input queue]
+</div>
+
+As soon as we deploy the new endpoint, it'll create a new subscription:
+
+<div class="mermaid">
+graph LR
+    P[Publisher] --> T[PaymentCompleted - SNS Topic]
+    T --> S1[Sales - subscription]
+    S1 -->Q1[Sales - SQS input queue]
+    T --> S2[Sales.PostPaymentProcessing - subscription]
+    S2 -->Q2[Sales.PostPaymentProcessing - SQS input queue]
+</div>
+
+The Sales endpoint receives the `PaymentCompleted` event and Sales.PostPaymentProcessing too. That will cause duplicate processing.
+
+As we discussed, we need to keep the message handlers in the Sales endpoint to process in-flight messages: those sitting in the input queue and those sent before we apply the changes. But we must avoid getting new messages.
+
+To achieve that, we need to delete the Sales subscription to the SNS topic and modify the Sales endpoint configuration to stop subscribing to the `PaymentCompleted` event while keeping the message handler.
+
+Note that if you use NServiceBus, it will recreate the deleted subscription every time the endpoint restarts. To prevent that, the [`AutoSubscribe` feature comes with the option to disable subscriptions for specific events](https://docs.particular.net/nservicebus/messaging/publish-subscribe/controlling-what-is-subscribed?version=core_8#automatic-subscriptions-exclude-event-types-from-auto-subscribe). Thanks, [Tim](https://timbussmann.github.io/), for the reminder.
```

Comment:
Twitter might be a better link for Tim, if he has one

Reply:
He has, but he also wants to be anonymous 😃